File: server/src/routes/preview.js
import express from 'express';
import axios from 'axios';

const router = express.Router();

const FASTAPI_URL = process.env.FASTAPI_URL || 'http://127.0.0.1:8000';

/**
 * POST /api/preview/pptx
 * Proxies the presigned S3 URL to FastAPI for LibreOffice conversion
 * and pipes the returned PDF buffer to the client.
 */
router.post('/pptx', async (req, res) => {
  try {
    const { url } = req.body;
    
    if (!url || typeof url !== 'string') {
      return res.status(400).json({ error: 'url is required in request body' });
    }

    // Forward to FastAPI requesting a stream back
    const response = await axios.post(`${FASTAPI_URL}/api/preview/pptx`, { url }, {
      responseType: 'stream',
      timeout: 60000
    });

    res.set({
      'Content-Type': 'application/pdf',
      'Content-Disposition': 'inline; filename="preview.pdf"',
    });

    // Pipe the PDF stream directly back to the client
    response.data.pipe(res);
  } catch (err) {
    if (err.response) {
      // Consume the error stream to get the JSON error message
      let errorData = '';
      err.response.data.on('data', chunk => errorData += chunk);
      err.response.data.on('end', () => {
        try {
          const parsed = JSON.parse(errorData);
          res.status(err.response.status).json(parsed);
        } catch {
          res.status(err.response.status).json({ error: 'Failed to convert PPTX to PDF' });
        }
      });
    } else {
      console.error('[Preview Proxy] Error connecting to FastAPI:', err.message);
      res.status(500).json({ error: 'Preview Microservice Unavailable' });
    }
  }
});

export default router;

File: fastapi-server/routers/preview.py
import os
import tempfile
import subprocess
import httpx
from fastapi import APIRouter, HTTPException
from fastapi.responses import Response
from pydantic import BaseModel

router = APIRouter()

MAX_PPTX_SIZE = 50 * 1024 * 1024

class PreviewRequest(BaseModel):
    url: str

@router.post("/pptx")
async def preview_pptx(req: PreviewRequest):
    if not req.url:
        raise HTTPException(status_code=400, detail="url is required in request body")
        
    async with httpx.AsyncClient() as client:
        try:
            resp = await client.get(req.url, timeout=30)
            resp.raise_for_status()
            buffer = resp.content
        except Exception:
            raise HTTPException(status_code=400, detail="Invalid file URL")
            
    if len(buffer) > MAX_PPTX_SIZE:
        raise HTTPException(status_code=400, detail="File too large for conversion")
        
    with tempfile.TemporaryDirectory(prefix="pptx-preview-") as tempdir:
        input_path = os.path.join(tempdir, "input.pptx")
        with open(input_path, "wb") as f:
            f.write(buffer)
            
        command = ["libreoffice", "--headless", "--convert-to", "pdf", "--outdir", tempdir, input_path]
        try:
            subprocess.run(command, check=True, timeout=45, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        except subprocess.CalledProcessError:
            raise HTTPException(status_code=500, detail="Failed to convert PPTX to PDF")
            
        pdf_path = os.path.join(tempdir, "input.pdf")
        if not os.path.exists(pdf_path):
            raise HTTPException(status_code=500, detail="Conversion succeeded but PDF not found")
            
        with open(pdf_path, "rb") as f:
            pdf_buffer = f.read()
            
    return Response(content=pdf_buffer, media_type="application/pdf", headers={
        "Content-Disposition": 'inline; filename="preview.pdf"'
    })
